pkg/core/interop_system.go

```diff
@@ -195,7 +195,13 @@ func storageFind(ic *interop.Context) error {
 	seekres := ic.DAO.SeekAsync(ctx, stc.ID, storage.SeekRange{Prefix: prefix})
 	item := istorage.NewIterator(seekres, prefix, opts)
 	ic.VM.Estack().PushItem(stackitem.NewInterop(item))
-	ic.RegisterCancelFunc(cancel)
+	ic.RegisterCancelFunc(func() {
+		cancel()
+		// Underlying persistent store is likely to be a private MemCachedStore. Thus,
+		// to avoid concurrent map iteration and map write we need to wait until internal
+		// seek goroutine is finished, because it can access underlying persistent store.
+		<-seekres
```

Comment:
Much better, but probably isn't sufficient. Your `cancel()` might be undelivered to `case <-ctx.Done()` reader when you'll get here and try to read from the `seekres` which will be satisfied by `case res <- KeyValue{Key: k, Value: v}`. Not very likely to happen, probably not an issue wrt to map access, but still we better read here until it's closed.

Reply:
Fixed. But maybe then better to use a separate goroutine responsible only for closing?